Configure ClusterSecretStore for infra cluster

larsks opened the pull request:
A ClusterSecretStore [1] creates a cluster-scoped secret store that will watch for ExternalSecret resources in all namespaces.

This commit creates a ClusterSecretStore that by defaults points at the vault instance on the local cluster and authenticates using a ServiceAccount. Every cluster overlay will need to patch spec.provider.vault.auth.kubernetes.mountPath, and clusters other than nerc-ocp-infra will need to (a) patch spec.provider.vault.server and (b) provide an alternative auth mechanism.

[1]: https://external-secrets.io/latest/api-clustersecretstore/

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202594685849610

larsks committed:
Configure ClusterSecretStore for infra cluster

A ClusterSecretStore [1] creates a cluster-scoped secret store that
will watch for ExternalSecret resources in all namespaces.

This commit creates a ClusterSecretStore that by defaults points at
the vault instance on the local cluster and authenticates using a
ServiceAccount.

[1]: https://external-secrets.io/latest/api-clustersecretstore/

dystewart commented on cluster-scope/base/core/serviceaccounts/vault-nerc-secret-reader/serviceaccount.yaml:
```diff
@@ -0,0 +1,5 @@
+kind: ServiceAccount
+metadata:
+  name: vault-nerc-secret-reader
+  namespace: external-secrets-config
```
In OPF we actually don't have a vault-secret-reader sa in the ESO namespace. We do have vault-secret-reader in each namespace which utilizes vault. We will need a token-reviewer in the  external-secrets-config namespace with auth delegtion crb:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: eso-tokenreview
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:auth-delegator
subjects:
  - kind: ServiceAccount
    name: eso-vault-auth
    namespace: external-secrets-config
```

dystewart replied:
That's not to say we can't do it this way though

larsks replied:
> We will need a token-reviewer in the external-secrets-config namespace with auth delegtion crb:

We don't need this on the cluster in which Vault is installed (at least, it all seems to work just fine without that). We will need both this RBAC and additional configuration for other clusters.

> In OPF we actually don't have a vault-secret-reader sa in the ESO namespace

For the `nerc-ocp-infra` cluster, I think it makes sense to just have a single `ClusterSecretStore` (in which case we just need a namespace to hold the `ServiceAccount`  used by the `ClusterSecretStore`). On the prod cluster we will probably want per-namespace `SecretStores` instead (to prevent us from accidentally exposing secrets to end-user namespaces).

dystewart replied:
Yeah agreed I think this makes sense. Also simplifies infra a bit and only admins should have access to this cluster anyway if I remember correctly. 
I also think in the prod cluster per namespace is the way to go.

larsks committed:
Move external secrets operator into dedicated namespace

Previously, we had the external secrets operator installed in the
openshift-operators namespace, but now that we're adding additional
configuration (the operatorconfig, service accounts, etc), it makes
sense to put it in its own namespace.

larsks committed:
Create ServiceAccount for Vault authentication

larsks committed:
Add clusterrolebindings for external secrets

Grant the  eso-vault-auth serviceaccount the system:auth-delegator
role, so that it can be used by Vault to validate tokens.

dystewart commented on cluster-scope/overlays/nerc-ocp-infra/clustersecretstores/nerc-cluster-secrets_patch.yaml:
```diff
@@ -0,0 +1,10 @@
+metadata:
+  name: nerc-cluster-secrets
+spec:
+  provider:
```
@larsks Oops I guess I didn't submit this comment yesterday. Is this configuration set up internally in Vault? If not I would be happy to get that started
